File: sdk/typescript/src/task.ts
/**
 * CapabilityCall + Task：async capability 调用句柄。镜像 Go task.go。
 * 链式 builder（withOnBehalfOfUser / withChainContext）+ invoke(sync) / submit(async)。
 */
import type { DispatcherClient } from "./dispatcher_client";
import type { EventsClient } from "./events_client";
import { BackendError, CancelledError, DispatcherRestartedError, TimeoutError } from "./errors";

const TERMINAL = new Set(["done", "failed", "cancelled"]);

export interface TaskInit {
  taskId: string;
  status: string;
  canonicalName: string;
  percent?: number;
  stageMessage?: string;
  resultPayload?: Record<string, unknown>;
  errorCode?: string;
  errorMessage?: string;
}

export class Task {
  taskId: string;
  status: string;
  canonicalName: string;
  percent: number;
  stageMessage: string;
  resultPayload: Record<string, unknown>;
  errorCode: string;
  errorMessage: string;

  constructor(
    init: TaskInit,
    private readonly dispatcher: DispatcherClient,
    // 注：构造参数属性名用 eventsClient（避免与下面的 events() 方法名冲突）。
    private readonly eventsClient?: EventsClient,
  ) {
    this.taskId = init.taskId;
    this.status = init.status;
    this.canonicalName = init.canonicalName;
    this.percent = init.percent ?? 0;
    this.stageMessage = init.stageMessage ?? "";
    this.resultPayload = init.resultPayload ?? {};
    this.errorCode = init.errorCode ?? "";
    this.errorMessage = init.errorMessage ?? "";
  }

  /** 主动拉一次快照覆盖本地字段。 */
  async refresh(opts: { signal?: AbortSignal } = {}): Promise<void> {
    const snap = await this.dispatcher.getTask(this.taskId, opts);
    this.status = snap.status;
    this.percent = snap.percent;
    this.stageMessage = snap.stageMessage;
    if (snap.result && Object.keys(snap.result).length > 0) this.resultPayload = snap.result;
    if (snap.errorCode) this.errorCode = snap.errorCode;
    if (snap.errorMessage) this.errorMessage = snap.errorMessage;
  }

  async cancel(opts: { signal?: AbortSignal } = {}): Promise<void> {
    await this.dispatcher.cancelTask(this.taskId, opts);
  }

  /** 等待终态。done→resultPayload；failed→DispatcherRestarted(50000)/BackendError；cancelled→Cancelled。 */
  async result(opts: { pollIntervalMs?: number; signal?: AbortSignal } = {}): Promise<Record<string, unknown>> {
    const interval = opts.pollIntervalMs ?? 2000;
    for (;;) {
      await this.refresh({ signal: opts.signal });
      if (TERMINAL.has(this.status)) break;
      if (opts.signal?.aborted) throw new TimeoutError("ctx cancelled");
      await sleep(interval, opts.signal);
    }
    switch (this.status) {
      case "done":
        return this.resultPayload;
      case "failed":
        if (this.errorCode === "50000" || this.errorCode === "DispatcherRestarted") {
          throw new DispatcherRestartedError(this.errorMessage);
        }
        throw new BackendError(this.errorMessage);
      case "cancelled":
        throw new CancelledError(this.errorMessage);
      default:
        throw new BackendError(`unexpected status=${this.status}`);
    }
  }

  /** lifecycle 事件流：注入 EventsClient → WS/polling；否则发一次 snapshot 后结束。 */
  events(): AsyncIterable<Record<string, unknown>> {
    if (this.eventsClient) {
      return this.eventsClient.register(this.taskId);
    }
    const self = this;
    return {
      async *[Symbol.asyncIterator]() {
        await self.refresh();
        yield {
          type: "snapshot", task_id: self.taskId, status: self.status,
          percent: self.percent, stage_message: self.stageMessage,
        };
      },
    };
  }
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    const id = setTimeout(resolve, ms);
    if (signal) {
      signal.addEventListener("abort", () => { clearTimeout(id); resolve(); }, { once: true });
    }
  });
}

/**
 * CapabilityCall：app.callCapability(name) 返的链式构造器。
 * .invoke(args) 走 sync；.submit(args) 走 async 拿 Task。
 * caller 引用他人能力写全名、不派生（不对称）。
 */
export class CapabilityCall {
  private onBehalfOfUserId = 0;
  private chainId = "";
  private chainHeader = "";

  constructor(
    public readonly canonicalName: string,
    private readonly dispatcher: DispatcherClient,
    private readonly events?: EventsClient,
  ) {}

  withOnBehalfOfUser(userId: number): this {
    this.onBehalfOfUserId = userId;
    return this;
  }

  withChainContext(chainId: string, chainHeader: string): this {
    this.chainId = chainId;
    this.chainHeader = chainHeader;
    return this;
  }

  async invoke(args: Record<string, unknown>, opts: { signal?: AbortSignal } = {}): Promise<Record<string, unknown>> {
    const res = await this.dispatcher.invoke({
      capability: this.canonicalName, args, mode: "sync",
      onBehalfOfUserId: this.onBehalfOfUserId, chainId: this.chainId, chainHeader: this.chainHeader,
      signal: opts.signal,
    });
    if (res.sync) return res.sync.result;
    throw new BackendError("expected sync but got async task_id");
  }

  async submit(args: Record<string, unknown>, opts: { signal?: AbortSignal } = {}): Promise<Task> {
    const res = await this.dispatcher.invoke({
      capability: this.canonicalName, args, mode: "async",
      onBehalfOfUserId: this.onBehalfOfUserId, chainId: this.chainId, chainHeader: this.chainHeader,
      signal: opts.signal,
    });
    if (res.async) {
      return new Task(
        { taskId: res.async.taskId, status: res.async.status, canonicalName: this.canonicalName },
        this.dispatcher, this.events,
      );
    }
    if (res.sync) {
      return new Task(
        { taskId: "", status: "done", canonicalName: this.canonicalName, resultPayload: res.sync.result },
        this.dispatcher, this.events,
      );
    }
    throw new BackendError("empty invoke result");
  }
}

File: sdk/go/ksapp/task.go
package ksapp

import (
	"context"
	"fmt"
	"time"
)

// Task 是 async capability 调用的句柄（Go 形态）。
//
// 与 Python `Task` 等价：caller 调 CallCapability(name).Submit(args) 拿到一个 Task，
// 然后 Refresh / Cancel / Result / Events 跟踪任务状态。
type Task struct {
	TaskID        string
	Status        string
	CanonicalName string
	Percent       int
	StageMessage  string
	ResultPayload map[string]any
	ErrorCode     string
	ErrorMessage  string

	dispatcher *DispatcherClient
	events     *EventsClient
}

var terminalStatuses = map[string]struct{}{
	"done": {}, "failed": {}, "cancelled": {},
}

// Refresh 主动拉一次最新快照覆盖本地字段。
func (t *Task) Refresh(ctx context.Context) error {
	snap, err := t.dispatcher.GetTask(ctx, t.TaskID)
	if err != nil {
		return err
	}
	t.Status = snap.Status
	t.Percent = snap.Percent
	t.StageMessage = snap.StageMessage
	if snap.Result != nil {
		t.ResultPayload = snap.Result
	}
	if snap.ErrorCode != "" {
		t.ErrorCode = snap.ErrorCode
	}
	if snap.ErrorMessage != "" {
		t.ErrorMessage = snap.ErrorMessage
	}
	return nil
}

// Cancel 调 keystone 取消任务。
func (t *Task) Cancel(ctx context.Context) error {
	return t.dispatcher.CancelTask(ctx, t.TaskID)
}

// Result 等待终态。
//   - done   → 返回 ResultPayload
//   - failed → 包装 ErrBackendError（或 ErrDispatcherRestarted）
//   - cancelled → ErrCancelled
//
// pollInterval=0 时取默认 2 秒；ctx 超时返 ErrTimeout 包装。
// EventsClient 接入后会优先用 event 驱动，poll 仅作 fallback。
func (t *Task) Result(ctx context.Context, pollInterval time.Duration) (map[string]any, error) {
	if pollInterval == 0 {
		pollInterval = 2 * time.Second
	}
	for {
		if err := t.Refresh(ctx); err != nil {
			return nil, err
		}
		if _, terminal := terminalStatuses[t.Status]; terminal {
			break
		}
		select {
		case <-ctx.Done():
			return nil, fmt.Errorf("%w: ctx cancelled", ErrTimeout)
		case <-time.After(pollInterval):
		}
	}
	switch t.Status {
	case "done":
		return t.ResultPayload, nil
	case "failed":
		if t.ErrorCode == "50000" || t.ErrorCode == "DispatcherRestarted" {
			return nil, fmt.Errorf("%w: %s", ErrDispatcherRestarted, t.ErrorMessage)
		}
		return nil, fmt.Errorf("%w: %s", ErrBackendError, t.ErrorMessage)
	case "cancelled":
		return nil, fmt.Errorf("%w: %s", ErrCancelled, t.ErrorMessage)
	}
	return nil, fmt.Errorf("%w: unexpected status=%s", ErrBackendError, t.Status)
}

// Events 返一个 channel 推送 lifecycle 事件。
// 已注入 EventsClient → 走 WS / polling；未注入则发一次 snapshot 后关闭。
func (t *Task) Events(ctx context.Context) (<-chan map[string]any, error) {
	if t.events != nil {
		stream := t.events.Register(t.TaskID)
		return stream.Channel(), nil
	}
	ch := make(chan map[string]any, 1)
	go func() {
		defer close(ch)
		_ = t.Refresh(ctx)
		select {
		case ch <- map[string]any{
			"type":          "snapshot",
			"task_id":       t.TaskID,
			"status":        t.Status,
			"percent":       t.Percent,
			"stage_message": t.StageMessage,
		}:
		case <-ctx.Done():
		}
	}()
	return ch, nil
}

// CapabilityCall 是 app.CallCapability(name) 返的构造器。
// 链式调 .Invoke(ctx, args) 走 sync；调 .Submit(ctx, args) 走 async 拿 *Task。
//
// 链式 options（v0.10.0+）：
//   - WithOnBehalfOfUser(uid)：透传调用链发起人 user_id，多跳 capability mesh
//     调用时必须设置（keystone dispatcher 端契约要求）。
//   - WithChainContext(chainID, chainHeader)：从上游 CapabilityContext 透传调用链，
//     让嵌套 app-to-app 调用继续落在同一 chain_id / chain_snapshot。
type CapabilityCall struct {
	canonicalName    string
	dispatcher       *DispatcherClient
	events           *EventsClient
	onBehalfOfUserID int64
	chainID          string
	chainHeader      string
}

// WithOnBehalfOfUser 设置调用链发起人 user_id，透传到 dispatcher 的
// on_behalf_of_user_id payload 字段，让 keystone 在多跳 capability mesh 调用中
// 维持初始 user 上下文。
//
// 典型使用：app-to-app 调用时，caller 应从 CapabilityContext.UserID() 取字符串
// 形态 user_id，atoi 后传入：
//
//	uid, _ := strconv.ParseInt(ctx.UserID(), 10, 64)
//	res, err := app.CallCapability("foo.bar").
//	    WithOnBehalfOfUser(uid).
//	    Invoke(reqCtx, args)
//
// userID=0 视为「无 user 上下文」（dispatcher payload 不带该字段，由 keystone
// 默认行为兜底，与 InvokeOptions.OnBehalfOfUserID=0 语义对齐）。
//
// 返回 *CapabilityCall 自身以便链式调用。
func (cc *CapabilityCall) WithOnBehalfOfUser(userID int64) *CapabilityCall {
	cc.onBehalfOfUserID = userID
	return cc
}

// WithChainContext 透传当前 capability 调用链上下文。
//
// 典型使用：在 capability handler 内嵌套调用其它 capability 时，把
// CapabilityContext.ChainID() 与 CapabilityContext.ChainHeader() 原样传入：
//
//	res, err := app.CallCapability("foo.bar").
//	    WithOnBehalfOfUser(uid).
//	    WithChainContext(ctx.ChainID(), ctx.ChainHeader()).
//	    Invoke(reqCtx, args)
//
// 任一参数为空时 SDK 不写对应 header，dispatcher 会按现有默认行为兜底。
func (cc *CapabilityCall) WithChainContext(chainID, chainHeader string) *CapabilityCall {
	cc.chainID = chainID
	cc.chainHeader = chainHeader
	return cc
}

// Invoke sync 调 dispatcher，返结果 map。
func (cc *CapabilityCall) Invoke(ctx context.Context, args map[string]any) (map[string]any, error) {
	res, err := cc.dispatcher.Invoke(ctx, InvokeOptions{
		Capability:       cc.canonicalName,
		Args:             args,
		Mode:             "sync",
		OnBehalfOfUserID: cc.onBehalfOfUserID,
		ChainID:          cc.chainID,
		ChainHeader:      cc.chainHeader,
	})
	if err != nil {
		return nil, err
	}
	if res.Sync != nil {
		return res.Sync.Result, nil
	}
	return nil, fmt.Errorf("%w: expected sync but got async task_id", ErrBackendError)
}

// Submit async 提交，返 *Task。
// 若 dispatcher 同步完成（mode=auto 时可能），返 Status="done" 已有 ResultPayload 的 Task。
func (cc *CapabilityCall) Submit(ctx context.Context, args map[string]any) (*Task, error) {
	res, err := cc.dispatcher.Invoke(ctx, InvokeOptions{
		Capability:       cc.canonicalName,
		Args:             args,
		Mode:             "async",
		OnBehalfOfUserID: cc.onBehalfOfUserID,
		ChainID:          cc.chainID,
		ChainHeader:      cc.chainHeader,
	})
	if err != nil {
		return nil, err
	}
	if res.Async != nil {
		return &Task{
			TaskID:        res.Async.TaskID,
			Status:        res.Async.Status,
			CanonicalName: cc.canonicalName,
			dispatcher:    cc.dispatcher,
			events:        cc.events,
		}, nil
	}
	if res.Sync != nil {
		return &Task{
			Status:        "done",
			CanonicalName: cc.canonicalName,
			ResultPayload: res.Sync.Result,
			dispatcher:    cc.dispatcher,
			events:        cc.events,
		}, nil
	}
	return nil, fmt.Errorf("%w: empty invoke result", ErrBackendError)
}
